Zend/zend_operators.c

```diff
@@ -596,24 +596,30 @@ ZEND_API void ZEND_FASTCALL convert_to_array(zval *op) /* {{{ */
 					HashTable *obj_ht = Z_OBJ_HT_P(op)->get_properties(op);
 					if (obj_ht) {
 						zval arr;
-
-						if (!Z_OBJCE_P(op)->default_properties_count && obj_ht == Z_OBJ_P(op)->properties) {
-							/* fast copy */
-							if (EXPECTED(Z_OBJ_P(op)->handlers == &std_object_handlers)) {
-								ZVAL_ARR(&arr, obj_ht);
-								if (EXPECTED(!(GC_FLAGS(Z_OBJ_P(op)->properties) & IS_ARRAY_IMMUTABLE))) {
-									GC_REFCOUNT(Z_OBJ_P(op)->properties)++;
-								}
+						zend_ulong num_key;
+						zend_string *str_key;
+						zval *zv;
+
+						/* Object property tables only have string keys, but PHP
+						 * arrays contain string and integer keys, and will
+						 * automatically coerce numeric strings to integers
+						 * when looking up values. Thus, we need to create a new
+						 * array with any numeric string keys converted into
+						 * integer keys, lest those values be inaccessible.
+						 */
+						array_init_size(&arr, zend_hash_num_elements(obj_ht));
+
+						ZEND_HASH_FOREACH_KEY_VAL(obj_ht, num_key, str_key, zv) {
+							Z_TRY_ADDREF_P(zv);
+							if (!str_key || ZEND_HANDLE_NUMERIC(str_key, num_key)) {
+								zend_hash_index_add_new(Z_ARRVAL(arr), num_key, zv);
 							} else {
-								ZVAL_ARR(&arr, zend_array_dup(obj_ht));
+								zend_hash_add_new(Z_ARRVAL(arr), str_key, zv);
 							}
-							zval_dtor(op);
-							ZVAL_COPY_VALUE(op, &arr);
-						} else {
-							ZVAL_ARR(&arr, zend_array_dup(obj_ht));
-							zval_dtor(op);
-							ZVAL_COPY_VALUE(op, &arr);
-						}
+						} ZEND_HASH_FOREACH_END();
+
+						zval_dtor(op);
+						ZVAL_COPY_VALUE(op, &arr);
```

Comment:
Here as well, this means that the non-copying fast-path is no longer present. The fast path applied primarily to casts of `stdClass` to arrays (as indirection prohibits its use for classes with declared properties).